core/container.go

```diff
@@ -339,6 +339,62 @@ func (container *Container) WithFS(ctx context.Context, dir *Directory) (*Contai
 	return &Container{ID: id}, nil
 }
 
+func (container *Container) WithDirectory(ctx context.Context, gw bkgw.Client, subdir string, src *Directory, filter CopyFilter) (*Container, error) {
+	dir, mount, err := locatePath(ctx, container, subdir, gw, NewDirectory)
+	if err != nil {
+		return nil, err
+	}
+
+	dir, err = dir.WithDirectory(ctx, ".", src, filter)
```

Comment:
Not 100% sure but think there might be a case needing handling still. Either way should have a test case for this if it's not already in there (didn't see it on a quick glance)

```go
mntDir := Directory().WithNewDirectory("/a/b/c")
ctr := Container().WithMountedDirectory("/mnt", mntDir)
newDir := Directory().WithNewDirectory("/foo")
ctr = ctr.WithDirectory("/a/b/foo", newDir)
// expected: "/mnt/a/b/foo"
// actual: "/mnt/foo" ??
```

Basically, I think when you grab `dir` from `locatePath` you end up getting a dir whose root is `/` (with the `/a/b` hidden), so when you remount it you end up dropping the preceding `/a/b`.

Again, not at all sure this is the case, just worth double checking in a test case as it's what I noticed reading the code.

Reply:
> // expected: "/mnt/a/b/foo"

I'm confused -- isn't this supposed to be `/a/b/foo/foo` given `ctr.WithDirectory("/a/b/foo", Directory().WithNewDirectory("/foo"))`?

Reply:
> I'm confused -- isn't this supposed to be /a/b/foo/foo given ctr.WithDirectory("/a/b/foo", Directory().WithNewDirectory("/foo"))?

Sorry, I had a typo in my example (even trying to write out an example is confusing 😞), I fixed it now, I meant to put it at `/mnt/a/b/foo`, not `/a/b/foo`

Reply:
Updated:

ctr.Exec("ls", "-al", "/mnt/a/b/foo") yields:

```
        	            	+total 12
        	            	+drwxr-xr-x    3 root     root          4096 Nov 17 19:40 .
        	            	+drwxr-xr-x    1 root     root          4096 Nov 17 19:40 ..
        	            	+drwxr-xr-x    2 root     root          4096 Nov 17 01:49 foo
```

That's good, right?

Reply:
> That's good, right?

Yeah that looks correct to me!

Reply:
The test is this:

```go
	mnt := c.Directory().WithNewDirectory("/a/b/c")
	ctr = c.Container().
		From("alpine:3.16.2").
		WithMountedDirectory("/mnt", mnt)
	dir = c.Directory().
		WithNewDirectory("/foo").
		WithNewFile("/foo/some-file", dagger.DirectoryWithNewFileOpts{
			Contents: "some-content",
		})
	ctr = ctr.WithDirectory("/mnt/a/b/foo", dir)
	contents, err = ctr.Exec(dagger.ContainerExecOpts{
		Args: []string{"cat", "/mnt/a/b/foo/foo/some-file"},
	}).Stdout().Contents(ctx)
	require.NoError(t, err)
	require.Equal(t, "some-content", contents)
```

I noticed that replacing `WithNewFile("/foo/some-file"` by `WithNewFile("some-file"` fails. I don't know if it's related to this PR or if relative `Directory.WithNewFile` are broken

Reply:
> I noticed that replacing WithNewFile("/foo/some-file" by WithNewFile("some-file" fails. I don't know if it's related to this PR or if relative Directory.WithNewFile are broken

Confirmed unrelated. Opened #3903

Reply:
Updated PR. I honestly don't know why CI is passing, but it is.